app/ide-desktop/lib/dashboard/src/authentication/src/authentication/config.ts

```diff
@@ -0,0 +1,136 @@
+ * connect to and use these pools. */
+
+import * as shared from '../../../../../client/shared'
+
+// =================
+// === Constants ===
+// =================
+
+/** AWS region in which our Cognito pool is located. */
+export const AWS_REGION: AwsRegion = 'eu-west-1' as AwsRegion
+/** URL used as the OAuth redirect when running in the desktop app. */
+export const DESKTOP_REDIRECT = `${shared.DEEP_LINK_SCHEME}://auth` as const
+/** Complete list of OAuth scopes used by the app. */
+export const OAUTH_SCOPES: OAuthScope[] = ['email' as OAuthScope, 'openid' as OAuthScope]
+/** OAuth response type used in the OAuth flows. */
+export const OAUTH_RESPONSE_TYPE: OAuthResponseType = 'code' as OAuthResponseType
+
+// =============
+// === Types ===
+// =============
+
+/** The AWS region in which our Cognito pool is located. This is always set to `eu-west-1` because
+ * that is the only region in which our Cognito pools are currently available in. */
+type AwsRegion = 'eu-west-1' & { _brand: 'AwsRegion' }
+/** ID of the "Cognito user pool" that contains authentication & identity data of our users.
+ *
+ * This is created automatically by our Terraform scripts when the backend infrastructure is
+ * created. Look in the `enso-org/cloud-v2` repo for details. */
+type UserPoolId = string & { _brand: 'UserPoolId' }
+/** ID of an OAuth client authorized to interact with the Cognito user pool specified by the
+ * {@link UserPoolId}.
+ *
+ * This is created automatically by our Terraform scripts when the backend infrastructure is
+ * created. Look in the `enso-org/cloud-v2` repo for details. */
+type UserPoolWebClientId = string & { _brand: 'UserPoolWebClientId' }
+/** Domain of the Cognito user pool used for authenticating/identifying the user.
+ *
+ * This must correspond to the public-facing domain name of the Cognito pool identified by the
+ * {@link UserPoolId}, and must not contain an HTTP scheme, or a pathname. */
+type OAuthDomain = string & { _brand: 'OAuthDomain' }
+/** Possible OAuth scopes to request from the federated identity provider during OAuth sign-in. */
+type OAuthScope = ('email' | 'openid') & { _brand: 'OAuthScope' }
+/** The response type used to complete the OAuth flow. "code" means that the federated identity
+ * provider will return an authorization code that can be exchanged for an access token. The
+ * authorization code will be provided as a query parameter of the redirect URL. */
+type OAuthResponseType = 'code' & { _brand: 'OAuthResponseType' }
+/** The URL used as a redirect (minus query parameters like `code` which get appended later), once
+ * an OAuth flow (e.g., sign-in or sign-out) has completed. These must match the values set in the
+ * Cognito pool and during the creation of the OAuth client. See the `enso-org/cloud-v2` repo for
+ * details. */
+export type OAuthRedirect = (
+    | typeof DESKTOP_REDIRECT
+    | 'http://localhost:8081'
```

Comment:
if you're not manipulating the string (and you're already using the brand for type-safety), it *might* be a better way to just do `string & { _brand: 'OAuthRedirect' }` as usual, and do `as OAuthRedirect` only when they're declared

Reply:
Ah, I see

Reply:
yeah - worth noting that closed types are plenty type-safe already. the reason why i think brands are more appropriate than constants here, is because it's the role those strings play that is important: "this string should be used for OAuth redirects", and the specific strings don't have any special meaning

(this is probably a lot easier to see when using them for ids like `ProjectId`)